src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,70 @@
+using GraphQL.Server.Transports.AspNetCore.Common;
+using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using GraphQLRequestBase = GraphQL.Server.Common.GraphQLRequest;
+
+namespace GraphQL.Server.Transports.AspNetCore.NewtonsoftJson
+{
+    /// <summary>
+    /// Implementation of an <see cref="IGraphQLRequestDeserializer"/> that uses Newtonsoft.Json.
+    /// </summary>
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializer _serializer;
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerSettings> configure)
+        {
+            var settings = new JsonSerializerSettings();
+            configure?.Invoke(settings);
+            _serializer = JsonSerializer.Create(settings); // it's thread safe https://stackoverflow.com/questions/36186276/is-the-json-net-jsonserializer-threadsafe
+        }
+
+        public Task<GraphQLRequestDeserializationResult> DeserializeFromJsonBodyAsync(HttpRequest httpRequest, CancellationToken cancellationToken = default)
+        {
+            // Do not explicitly or implicitly (via using, etc.) call dispose because StreamReader will dispose inner stream.
+            // This leads to the inability to use the stream further by other consumers/middlewares of the request processing
+            // pipeline. In fact, it is absolutely not dangerous not to dispose StreamReader as it does not perform any useful
+            // work except for the disposing inner stream.
+            var reader = new StreamReader(httpRequest.Body);
+
+            var result = new GraphQLRequestDeserializationResult() { IsSuccessful = true };
+
+            using (var jsonReader = new JsonTextReader(reader) { CloseInput = false })
+            {
+                switch (reader.Peek())
+                {
```

Comment:
Is it worth checking the cancellation token after the peek and before the full deserialization?

Reply:
Well, yes, but I'm sure that in fact it is not so principled.